uniffi_bindgen/src/bindings/kotlin/gen_kotlin.rs

```diff
@@ -63,236 +71,179 @@ pub struct KotlinWrapper<'a> {
-                "write{}({}, {})",
-                class_name_kt(&type_.canonical_name())?,
-                nm,
-                target,
-            ),
-            _ => format!("{}.write({})", nm, target),
-        })
+        let oracle = oracle();
+        Ok(oracle.find(type_)?.write(&oracle, nm, target))
     }
 
-    /// Get a Kotlin expression for lifting a value from something we received over the FFI.
-    ///
-    /// Where possible, this delegates to a `lift()` method on the type itself, but special
-    /// handling is required for some compound data types.
     pub fn lift_kt(nm: &dyn fmt::Display, type_: &Type) -> Result<String, askama::Error> {
-        let nm = nm.to_string();
-        Ok(match type_ {
-            Type::CallbackInterface(_) => format!(
-                "{}Internals.lift({})",
-                class_name_kt(&type_.canonical_name())?,
-                nm,
-            ),
-            Type::Optional(_)
-            | Type::Sequence(_)
-            | Type::Map(_)
-            | Type::Timestamp
-            | Type::Duration => format!("lift{}({})", class_name_kt(&type_.canonical_name())?, nm),
-            _ => format!("{}.lift({})", type_kt(type_)?, nm),
-        })
+        let oracle = oracle();
+        Ok(oracle.find(type_)?.lift(&oracle, nm))
+    }
+
+    pub fn literal_kt(literal: &Literal) -> Result<String, askama::Error> {
+        let type_ = match literal {
+            Literal::Enum(_, type_) => type_,
+            Literal::Int(_, _, type_) => type_,
+            Literal::UInt(_, _, type_) => type_,
+            Literal::Float(_, type_) => type_,
+            _ => return legacy_kt::literal_kt(literal),
+        };
+
+        let oracle = oracle();
+        Ok(oracle.find(type_)?.literal(&oracle, literal))
     }
 
-    /// Get a Kotlin expression for reading a value from a byte buffer.
-    ///
-    /// Where possible, this delegates to a `read()` method on the type itself, but special
-    /// handling is required for some compound data types.
     pub fn read_kt(nm: &dyn fmt::Display, type_: &Type) -> Result<String, askama::Error> {
-        let nm = nm.to_string();
-        Ok(match type_ {
-            Type::CallbackInterface(_) => format!(
-                "{}Internals.read({})",
-                class_name_kt(&type_.canonical_name())?,
-                nm,
-            ),
-            Type::Optional(_)
-            | Type::Sequence(_)
-            | Type::Map(_)
-            | Type::Timestamp
-            | Type::Duration => format!("read{}({})", class_name_kt(&type_.canonical_name())?, nm),
-            _ => format!("{}.read({})", type_kt(type_)?, nm),
-        })
+        let oracle = oracle();
+        Ok(oracle.find(type_)?.read(&oracle, nm))
     }
-}
+
+    /// Get the Kotlin syntax for representing a given low-level `FFIType`.
+    pub fn type_ffi(type_: &FFIType) -> Result<String, askama::Error> {
+        Ok(oracle().ffi_type_label(type_))
+    }
```

Comment:
To check my expectations - the idea is that these filters will eventually go away in favour of directly calling methods on the `CodeType`, right?

Reply:
I think so. If there was an easier way to get things into a template, without making it part of a struct this would be my goal.

Reply:
The filters are all now incredibly regular. I'm resisting the urge to do away with filters yet, or pass the `oracle` into the filters from the `MemberDeclaration`.